lib/spack/spack/package_prefs.py

```diff
@@ -167,24 +167,27 @@ def spec_externals(spec):
     external_specs = []
     for name in names:
         pkg_config = allpkgs.get(name, {})
-        pkg_paths = pkg_config.get('paths', {})
-        pkg_modules = pkg_config.get('modules', {})
-        if (not pkg_paths) and (not pkg_modules):
-            continue
-
-        for external_spec, path in pkg_paths.items():
-            external_spec = spack.spec.Spec(
-                external_spec, external_path=canonicalize_path(path))
-            if external_spec.satisfies(spec):
-                external_specs.append(external_spec)
-
-        for external_spec, module in pkg_modules.items():
-            external_spec = spack.spec.Spec(
-                external_spec, external_module=module)
+        pkg_externals = pkg_config.get('externals', [])
+        for entry in pkg_externals:
+            # This copy makes it safe to pop out of entry without
+            # modifying the object in config
+            entry = copy.deepcopy(entry)
+            spec_str = entry.pop('spec')
+            external_path = entry.pop('prefix', None)
+            if external_path:
+                external_path = canonicalize_path(external_path)
+            external_module = entry.pop('module', None)
+            external_spec = spack.spec.Spec.from_detection(
+                spack.spec.Spec(
+                    spec_str,
+                    external_path=external_path,
+                    external_module=external_module
+                ), extra_attributes=entry
```

Comment:
It looks to me like entry is a list item of `externals`, which is a dictionary that looks like

```
{
  spec: ...
  module: ...
  path: ...
  extra_attributes: {}
}
```

You are popping `spec`, `module`, etc. leaving just `extra_attributes`, but shouldn't this still set `extra_attributes=entry['extra_attributes']`? Otherwise you'd have

`extra_attributes={'extra_attributes': {the extra attributes}}`

rather than

`extra_attributes={the extra attributes}`.

Also, is there a reason to do a deepcopy other than to pop items off and be left with extra_attributes by itself? Generally I think it's best to avoid deepcopy because the semantics can be unexpected depending on the implementation of the object being copied.

Reply:
> It looks to me like entry is a list [ ... ] 

The dictionary looks like https://github.com/spack/spack/pull/16526#issuecomment-628919889 the extra attributes are added directly in `packages.yaml`, not under the `extra_attributes` key.

> Also, is there a reason to do a deepcopy other than to pop items off and be left with extra_attributes by itself?

No, I did it to be defensive and avoid changing anything that could be referenced by the configuration. Changed to `copy.copy`.

Reply:
> The dictionary looks like #16526 (comment) the extra attributes are added directly in packages.yaml, not under the extra_attributes key.

I missed that - thanks for the pointer. I prefer a "clean" namespace but this is consistent with other Spack configuration.

Also, I could see deepcopy being useful, but I'm wondering if it might make sense to wrap it in a function offered by the `spack_yaml` objects themselves. This becomes difficult because in some cases they are storing `ruamel.yaml` objects. It may be useful to use something like `_convert_to_nested_dict` from https://github.com/spack/spack/pull/15939.

This all being said, based on the version update info from https://pypi.org/project/ruamel.yaml/, problems with deepcopy for ruamel.yaml objects have been addressed so this shouldn't be a problem (although IMO it's still worth avoiding deepcopy to avoid having to do this type of verification).